In [13]:
import sklearn as sk
from sklearn import preprocessing
import math
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
import mglearn
import os
import inspect
import sys


sys.path.append('./../../')

# OPTIONAL: Load the "autoreload" extension so that code can change
%load_ext autoreload

#Own Library modules
import src.pca

# OPTIONAL: always reload modules so that as you change code in src, it gets loaded
%autoreload 2

from src.pca import PCA_from_sklearn

# Importamos la función componentes_principales del archivo .py
#import PCA_from_sklearn

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


#### Objetivo:

El objetivo de este notebook es utilizar PCA () de scikit-learn para la reducción de dimensionalidad por sección, ya que existen más de 150 variables


In [2]:
df = pd.read_csv('datos_limpios.csv', encoding = "L1")

##### Sección 3. Expectativas educativas

In [3]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_3 = df.loc[:, ['p10', 'p11', 'p12', 'p13', 'p14']] ###Se seleccionan las preguntas que conforman esta sección

In [4]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_3 = pd.DataFrame(scaler.fit_transform(seccion_3))

In [5]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_3)

In [6]:
componentes = pd.DataFrame(pca.components_.T, index=['p10', 'p11', 
                                                         'p12', 'p13',
                                                         'p14', 
                                                         ])

In [7]:
####Se eligieron dos componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos 
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2'})
componentes

principal_component_1  principal_component_2
p10               0.118769              -0.745795
p11               0.433563               0.450160
p12               0.487973               0.328743
p13               0.530490              -0.199727
p14               0.527617              -0.305260

In [8]:
####Se obtiene la varianza explicada por cada componente principal
varianza_explicada

array([0.49697126, 0.21837784])

Por lo tanto, las preguntas 13 y 14 se encuentran asociadas al primer componente principal a un porcentaje de varianza explicada de 49.70%, mientras que la pregunta 11 se asocia al componente principal 2 a un porcentaje de varianza explicada de 21.84%.

- Como se observa los eigenvalues son mayores a 1, siendo está otra medida para escoger el número de componentes, reduciendo así la dimensionalidad.

In [9]:
eigenvalues

array([2.48486074, 1.09189113])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [10]:
componentesprincipales_seccion_3 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_3

0         1
0      -2.751152  1.038033
1       1.439305 -0.274336
2      -0.075647 -1.163561
3      -2.054591  0.706529
4      -0.592569  1.411202
...          ...       ...
560621 -1.369152  0.309960
560622 -1.242848 -0.348250
560623  1.372021 -1.269330
560624 -1.481507 -0.938306
560625 -0.546287 -0.679755

[560626 rows x 2 columns]

##### Sección 4. Características personales

- 4.1. Indica qué tanto te describe lo siguiente.

In [11]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados
seccion_4_1 = df.loc[:, ['p15', 'p16', 'p17', 'p18', 'p19', 'p20', 'p21', 'p22']] ###Se seleccionan las preguntas que conforman esta sección

In [12]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_4_1 = pd.DataFrame(scaler.fit_transform(seccion_4_1))

In [13]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_4_1)

In [14]:
componentes = pd.DataFrame(pca.components_.T, index=['p15', 'p16', 'p17', 'p18', 'p19', 'p20', 'p21', 'p22'])

In [15]:
####Se eligieron tres componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3'})
componentes

principal_component_1  principal_component_2  principal_component_3
p15               0.264701               0.378767               0.105304
p16              -0.173163               0.300260               0.846667
p17               0.346735               0.324078               0.225066
p18              -0.440740               0.307729              -0.070533
p19               0.203013               0.466745              -0.340611
p20               0.294675               0.451538              -0.230419
p21              -0.487734               0.219685              -0.144270
p22              -0.468550               0.310898              -0.162831

In [16]:
####Se obtiene la varianza explicada por cada componente principal
varianza_explicada

array([0.28505091, 0.24167855, 0.11123021])

La pregunta 17 se asocia al primer componente principal, la pregunta 19 y la pregunta 20 se encuentran asociadas al segundo componente principal y la pregunta 16 se asocia al tercer componente principal. La varianza explicada para estos componentes es de 28.50%, 24.17% y 11.12% respectivamente.

- Aquí se puede observar que no todos los eigenvalues son mayores a 1, por lo que esta medida nos dice que nos quedemos con 2 componentes, sin embargo, la media que usamos es la varianza explicada.

In [17]:
eigenvalues

array([2.28041137, 1.93343182, 0.88984326])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [18]:
componentesprincipales_seccion_4_1 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_4_1

0         1         2
0      -0.030844 -0.040263  0.652843
1       0.605189  0.816913  2.242497
2      -0.737430  1.031727  1.150211
3       0.832951  0.426113  0.362032
4       1.640749 -0.793292  0.279120
...          ...       ...       ...
560621  1.403816 -0.980054 -0.049580
560622  1.117009 -0.480046 -0.389970
560623 -2.038386  1.683265 -0.697210
560624  0.697905  2.291276 -0.621842
560625 -0.699533  2.572282  1.000256

[560626 rows x 3 columns]

- 4.2. Indica tu grado de acuerdo con los siguientes enunciados.

In [19]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_4_2 = df.loc[:, ['p23', 'p24', 'p25', 'p26', 'p27', 'p28', 'p29', 'p30', 'p31', 'p32', 'p33']] ###Se seleccionan las preguntas que conforman esta sección

In [20]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_4_2 = pd.DataFrame(scaler.fit_transform(seccion_4_2))

In [21]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_4_2)

In [22]:
componentes = pd.DataFrame(pca.components_.T, index=['p23', 'p24', 'p25', 'p26', 'p27', 'p28', 'p29', 'p30', 'p31', 'p32', 'p33'])

In [23]:
####Se eligieron cinco componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 
                                         3:'principal_component_4', 4:'principal_component_5'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p23              -0.371245              -0.141233              -0.261124   
p24              -0.375638               0.010097              -0.359534   
p25              -0.385254               0.031232              -0.083717   
p26              -0.335045              -0.064105              -0.275946   
p27              -0.376515               0.083793              -0.102070   
p28              -0.232810              -0.073033               0.341450   
p29              -0.336734              -0.060097               0.216034   
p30               0.002200               0.675097              -0.094035   
p31              -0.324900               0.010834               0.338898   
p32              -0.032640               0.684145              -0.091101   
p33              -0.210743               0.187007               0.643213   

     principal_component_4  principal_component_5  
p23              -0.094564              -0.040953  
p24               0.009110              -0.094674  
p25               0.246011               0.256487  
p26              -0.067090              -0.213339  
p27               0.053251              -0.111202  
p28              -0.856636               0.180779  
p29               0.229667               0.574780  
p30              -0.057461               0.465976  
p31              -0.012301              -0.330252  
p32              -0.149085              -0.344068  
p33               0.332963              -0.240954

In [24]:
####Se obtiene la varianza explicada por cada componente principal
varianza_explicada

array([0.25432501, 0.11525382, 0.09639745, 0.08056387, 0.07513842])

Por lo tanto, la pregunta 30 se asocia al primer componente principal, el cual para este componente su varianza explicada es de 25.43%. Las preguntas 30 y 32 se asocian al segundo componente principal con un porcentaje de varianza explicada de 11.52%. La pregunta 33 se asocia al tercer componente principal con un 9.64% de varianza explicada y también al cuarto componente principal con una varianza explicada de 8.05%; mientras que la pregunta 29 se asocia al quinto componente principal con un porcentaje de varianza explicada de 7.51%.  

- Los eigenvalues nos muestran que no todos lo componentes que se eligieron considerando la varianza explicada son mayores a 1.

In [25]:
eigenvalues

array([2.79758012, 1.26779433, 1.06037382, 0.8862042 , 0.82652405])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [26]:
componentesprincipales_seccion_4_2 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_4_2

0         1         2         3         4
0       2.654835 -0.126711  0.992496 -0.340768  1.132157
1       0.244777 -0.627959 -2.188335 -0.731437  0.291323
2       0.366210  0.129824  0.785720  1.329208  0.227957
3       2.503022 -1.133014 -1.519482  0.265603 -0.282153
4       0.979646  0.587305 -0.811460 -0.286928  0.013899
...          ...       ...       ...       ...       ...
560621  0.599428 -0.152860 -1.541828  0.538943 -1.032911
560622  2.119818  0.758164  0.625487 -0.482713 -0.541124
560623 -2.765005 -1.735203 -1.077220  1.814559  0.136786
560624 -2.023575 -1.555157  1.493046 -0.153962  0.198163
560625 -1.425262  0.527075 -0.957056 -0.378311 -1.178684

[560626 rows x 5 columns]

- 4.3. En el último mes, ¿con qué frecuencia te sucedió lo siguiente?

In [27]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_4_3 = df.loc[:, ['p34', 'p35', 'p36', 'p37', 'p38', 'p39', 'p40']
                  ] ###Se seleccionan las preguntas que conforman esta sección

In [28]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_4_3 = pd.DataFrame(scaler.fit_transform(seccion_4_3))

In [29]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_4_3)

In [30]:
componentes = pd.DataFrame(pca.components_.T, index=['p34', 'p35', 'p36', 'p37', 'p38', 'p39', 'p40'])

In [31]:
####Se eligieron tres componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3'})
componentes

principal_component_1  principal_component_2  principal_component_3
p34               0.385154              -0.035362              -0.205920
p35               0.416554              -0.256862              -0.209465
p36               0.407836              -0.307341              -0.114798
p37               0.413473              -0.131209               0.021778
p38               0.415866              -0.122336               0.287020
p39               0.318163               0.496189               0.715731
p40               0.258218               0.748283              -0.552646

In [32]:
####Se obtiene la varianza explicada por cada componente principal
varianza_explicada

array([0.4473773 , 0.13501357, 0.10271715])

Las preguntas 35, 37 y 38 se asocial al primer componente principal, el cual para este componente le corresponde una varianza explicada de 44.74%. La pregunta 40 se asocia al segundo componente principal a un porcentaje de varianza explicada de 13.50%, mientras que la pregunta 39 se asocia al tercer componente principal. A este último le corresponde una varianza explicada de 10.27%.

- Se observa que el eigenvalue asociado al primer componente principal es mayor a 1. 

In [33]:
eigenvalues

array([3.13164666, 0.94509666, 0.71902136])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [34]:
componentesprincipales_seccion_4_3 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_4_3

0         1         2
0      -0.281886 -0.954739  0.381717
1      -1.508087  0.984857 -0.438893
2       1.129132  0.439207 -1.480781
3      -1.555394 -0.935657  0.570194
4       0.263456  0.716205 -0.014434
...          ...       ...       ...
560621  2.723442 -0.517490  1.083444
560622 -0.503595  0.086827  0.214320
560623 -2.464391  1.173674 -0.539030
560624  2.345588 -0.731864  0.196930
560625  1.790942  1.846238 -0.028493

[560626 rows x 3 columns]

- 4.4. Cuando tienes que tomar una decisión, ¿con qué frecuencia realizas lo siguiente? 

In [35]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_4_4 = df.loc[:, ['p41', 'p42', 'p43', 'p44', 'p45', 'p46', 'p47', 'p48', 'p49', 'p50',
                      'p51', 'p52', 'p53', 'p54', 'p55', 'p56']] ###Se seleccionan las preguntas que conforman esta sección

In [36]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_4_4 = pd.DataFrame(scaler.fit_transform(seccion_4_4))

In [37]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_4_4)

In [38]:
componentes = pd.DataFrame(pca.components_.T, index=['p41', 'p42', 'p43', 'p44', 'p45', 'p46', 'p47', 'p48', 'p49', 'p50',
                      'p51', 'p52', 'p53', 'p54', 'p55', 'p56'])

In [39]:
####Se eligieron seis componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 
                                         3:'principal_component_4', 4:'principal_component_5', 5:'principal_component_6'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p41              -0.093297              -0.484718               0.012643   
p42              -0.077733              -0.507544               0.007779   
p43              -0.082770              -0.474760               0.022953   
p44              -0.117975              -0.445595              -0.034594   
p45               0.135970              -0.196407               0.144808   
p46               0.271874              -0.022436               0.482993   
p47               0.288919              -0.025086               0.263688   
p48               0.255210              -0.073081               0.487855   
p49               0.316425              -0.059288               0.168224   
p50               0.322587              -0.081334              -0.028065   
p51               0.345181               0.017972              -0.024412   
p52               0.264506               0.000847              -0.188579   
p53               0.333248              -0.063539              -0.279881   
p54               0.320647              -0.049802              -0.319190   
p55               0.303126              -0.073655              -0.311945   
p56               0.156269              -0.126684              -0.305778   

     principal_component_4  principal_component_5  principal_component_6  
p41               0.183358              -0.002645              -0.041707  
p42               0.156653              -0.049447              -0.025999  
p43               0.054580              -0.121101              -0.016339  
p44              -0.136049              -0.030603               0.033303  
p45              -0.847575              -0.034953               0.189708  
p46              -0.015845               0.132269              -0.285033  
p47              -0.164243               0.022856              -0.109927  
p48               0.195977               0.126050              -0.280023  
p49               0.269850               0.018583               0.366563  
p50               0.145915              -0.091086               0.550718  
p51               0.095464              -0.092499               0.364064  
p52              -0.161200              -0.268688              -0.110768  
p53               0.059511              -0.213005              -0.281006  
p54               0.020611              -0.170789              -0.319474  
p55              -0.020824               0.088994              -0.155171  
p56              -0.052205               0.880322               0.030036

In [40]:
varianza_explicada

array([0.25370117, 0.15133278, 0.07891853, 0.05810544, 0.05522506,
       0.05009503])

Para el primer componente principal se asocian las preguntas 51 y 53, la pregunta 51 se asocia al segundo componente principal, la pregunta 46 y 48 se asocial al tercer componente principal, la pregunta 49 se asocia al cuarto componente principal, la pregunta 56 se asocia al quinto componente principal y por último, la pregunta 50 se asocia al sexto componente principal. La varianza explicada para cada uno de los componentes principales es de 25.37%, 15.13%, 7.89%, 5.81%, 5.52% y 5.01%, respectivamente.

- A continuación se observa que únicamente 3 eigenvalues son mayores a 1.

In [41]:
eigenvalues

array([4.05922595, 2.42132877, 1.26269872, 0.92968875, 0.88360258,
       0.80152195])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [42]:
componentesprincipales_seccion_4_4 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_4_4

0         1         2         3         4         5
0      -0.751822 -0.670241 -0.712069 -0.392792 -0.014749 -0.624745
1      -0.018447 -0.246731  0.443421 -0.358973 -0.420004  0.503786
2       0.637087  1.068688  0.809777  0.075249 -0.299770 -1.898115
3      -0.579509  1.381914  1.186150  0.820790  0.421591  0.235735
4       0.590726 -0.127315  0.604520  0.661742 -0.451740 -0.434929
...          ...       ...       ...       ...       ...       ...
560621  2.729790  1.187233  0.723093 -0.940091 -0.496617 -1.257455
560622  0.281670  1.509833 -0.115315  0.135614  0.795303  0.666997
560623 -0.842480 -1.880479  1.798957  1.343685 -0.847550 -1.893355
560624  0.030541  0.822774 -0.904205 -0.614341  0.800378  1.942547
560625 -1.184392 -1.002625  2.217696  1.246987 -0.830297  0.444196

[560626 rows x 6 columns]

##### Sección 5. Inglés y cómputo.

- 5.1. Califica tu nivel de habilidad para leer en inglés.

In [43]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_5_1 = df.loc[:, ['p57', 'p58', 'p59', 'p60']] ###Se seleccionan las preguntas que conforman esta sección

In [44]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_5_1 = pd.DataFrame(scaler.fit_transform(seccion_5_1))

In [45]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_5_1)

In [46]:
componentes = pd.DataFrame(pca.components_.T, index=['p57', 'p58', 'p59', 'p60'])

In [47]:
####Se eligieron un componente principal, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2'})
componentes

principal_component_1
p57               0.506117
p58               0.517654
p59               0.488779
p60               0.486801

In [48]:
varianza_explicada

array([0.74731825])

Para el primer componente principal se asocian la preguntas 57 y 58. El porcentaje de varianza explicada para el primer componentes es de 74.73%.

- El eigenvalue asociado a este componente principal es mayor a 1.

In [49]:
eigenvalues

array([2.98927833])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [50]:
componentesprincipales_seccion_5_1 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_5_1

0
0      -3.012403
1       1.329468
2      -1.121389
3      -1.763936
4      -0.536424
...          ...
560621 -2.402316
560622 -0.536424
560623  3.805447
560624  1.967849
560625 -2.402316

[560626 rows x 1 columns]

- 5.2. Con qué frecuencia realizas las siguientes actividades en la computadora

In [51]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_5_2 = df.loc[:, ['p61', 'p62', 'p63', 'p64']] ###Se seleccionan las preguntas que conforman esta sección

In [52]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_5_2 = pd.DataFrame(scaler.fit_transform(seccion_5_2))

In [53]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_5_2)

In [54]:
componentes = pd.DataFrame(pca.components_.T, index=['p61', 'p62', 'p63', 'p64'])

In [55]:
####Se eligieron dos componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2'})
componentes

principal_component_1  principal_component_2
p61               0.450106              -0.719827
p62               0.548978              -0.023552
p63               0.536229               0.038504
p64               0.456602               0.692684

In [56]:
varianza_explicada

array([0.59243727, 0.19244308])

Para el primer componente principal se asocia la pregunta 62 y 63, mientras que para el segundo componente principal se asocia la pregunta 64. El porcentaje de varianza explicada para el primer componentes es de 59.24% y para el segundo componente principal es de 19.24%.

- Se observa que únicamente el primer eigenvalue asociado al primer componente principal es mayor a 1.

In [57]:
eigenvalues

array([2.36975333, 0.76977369])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [58]:
componentesprincipales_seccion_5_2 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_5_2

0         1
0      -0.686659 -1.686094
1      -0.710620 -0.395172
2      -0.596293 -1.027102
3      -0.696755  0.317548
4      -2.404186  0.325751
...          ...       ...
560621 -1.461989 -0.377540
560622 -1.022208 -0.345961
560623 -0.596293 -1.027102
560624  0.785685 -1.698813
560625  0.245442 -0.385742

[560626 rows x 2 columns]

##### Sección 7. Trayectoria académica

- 7.1. Primaria y secundaria

In [59]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_7_1 = df.loc[:, ['p67', 'p68', 'p69', 'p70', 'p71']] ###Se seleccionan las preguntas que conforman esta sección

In [60]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_7_1 = pd.DataFrame(scaler.fit_transform(seccion_7_1))

In [61]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_7_1)

In [62]:
componentes = pd.DataFrame(pca.components_.T, index=['p67', 'p68', 'p69', 'p70', 'p71'])

In [63]:
####Se eligieron tres componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3'})
componentes

principal_component_1  principal_component_2  principal_component_3
p67               0.045094               0.748319               0.259472
p68              -0.665152              -0.071458               0.201885
p69               0.241351              -0.647058               0.378957
p70               0.298964               0.108741               0.801152
p71              -0.638678              -0.066361               0.326289

In [64]:
varianza_explicada

array([0.33720904, 0.2575904 , 0.21631804])

Para el primer componente principal se asocian las preguntas 69 y 70, para el segundo componente principal se asocia la pregunta 67 y para el tercer componente principal se asocia la pregunta 70. La varianza explicada para cada uno de los componentes principales es de 33.72%, 25.76% y 21.63%, respectivamente.

- Se observa que los tres eigenvalues asociados a los componentes principales son mayores a 1.

In [65]:
eigenvalues

array([1.68604821, 1.28795431, 1.08159214])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [66]:
componentesprincipales_seccion_7_1 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_7_1

0         1         2
0       0.409965 -0.826329 -0.947670
1       2.012770 -0.580627 -0.723277
2      -0.593213 -0.855984  0.338505
3       2.147287 -0.642757 -1.655524
4      -0.458696 -0.918114 -0.593742
...          ...       ...       ...
560621 -0.294408 -0.747300  1.139232
560622  0.138915 -0.700748  1.007711
560623 -1.987769  2.463065 -0.907243
560624 -0.460711 -0.916795 -0.502857
560625  2.193462  0.088737 -1.492313

[560626 rows x 3 columns]

- 7.2. Educación media superior (Bachillerato)

In [67]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_7_2 = df.loc[:, ['p72', 'p73', 'p74', 'p75', 'p76', 'p77', 'p78']] ###Se seleccionan las preguntas que conforman esta sección

In [68]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_7_2 = pd.DataFrame(scaler.fit_transform(seccion_7_2))

In [69]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_7_2)